drivers/can/socket_can_generic.h

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2019 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+/* Generic CANBUS related functions that are generic in all the drivers. */
+
+#include <net/net_pkt.h>
+#include <net/socket_can.h>
+
+#ifndef ZEPHYR_DRIVERS_CAN_SOCKET_CAN_GENERIC_H_
+#define ZEPHYR_DRIVERS_CAN_SOCKET_CAN_GENERIC_H_
+
+#define SOCKET_CAN_NAME_1 "SOCKET_CAN_1"
+#define SEND_TIMEOUT K_MSEC(100)
+#define RX_THREAD_STACK_SIZE 512
+#define RX_THREAD_PRIORITY 2
+#define BUF_ALLOC_TIMEOUT K_MSEC(50)
+
+/* TODO: make msgq size configurable */
+CAN_DEFINE_MSGQ(socket_can_msgq, 5);
+K_THREAD_STACK_DEFINE(rx_thread_stack, RX_THREAD_STACK_SIZE);
+
+struct socket_can_context {
+	struct device *can_dev;
+	struct net_if *iface;
+	struct k_msgq *msgq;
```

Comment:
So, here's exactly what I mean - for CAN, you have your own queues and infrastructure anyway. Sockets could work with this infrastructure directly. Instead, `net/ip/` code was slashed and twisted to useless plug all this unrelated CAN stuff into it.

Reply:
It would actually be great when all raw sockets can bypass the IP stack somehow. This would reduce the memory footprint a lot.

Reply:
> when all raw sockets can bypass the IP stack somehow.

Well, "all raw sockets" can't bypass IP stack, because there're IP raw sockets. But sockets of address families not related to IP (e.g., AF_PACKET, AF_CAN) definitely can bypass IP stack, and it would make the IP stack much more manageable (given that we discover bug after bug in it).

Reply:
Regarding the msgq: As already discussed above, this will be removed and the isr will be used instead. For now we have two threads that pass the same massage.